### PR TITLE
Feature/867 command palette

### LINKED ISSUE #867: Command Palette (Cmd/Ctrl+K)
> ℹ️ _This issue was partially AI-generated and may contain technical inaccuracies._

**Is your feature request related to a problem? Please describe.**
Navigating to a specific request or triggering an action requires mouse-driven sidebar browsing, which is slow in large collections.

**Describe the solution you'd like**
Add a command palette (Cmd/Ctrl+K) for fuzzy-searching requests, folders, and collections, and for running common actions (send, save, new request, switch environment, etc.). The `cmdk` dependency is already in the project.

**Describe alternatives you've considered**
Hotkeys for individual actions (#446); the sidebar search.

**Additional context**
Distinct from #446 (single-key hotkeys). Reuses the existing `cmdk` library and complements the keyboard-driven workflow.

### PR DESCRIPTION
**Pull request by eugen-soliar**
## Changes
For #867 

- Added Command Palette (mod+k) with All / Requests / Environments / Actions tabs
- Requests tab groups by folder; method badges aligned via CSS subgrid
- All tab: "Recent" (5 most-recent) on empty search, cross-section search otherwise
- Environments tab with active-env indicator
- Actions tab: Send/Save/New/Discard request, New Folder, Switch Environment, Collection Settings, Trufos Settings
- Migrated AppSettingsModal to controlled viewStore state
- Platform-aware shortcut badges (⌘ on Mac, Ctrl elsewhere)
- Selecting a request now expands ancestor folders and scrolls it into view in the sidebar
- Fixed hover/keyboard highlight color regression (divider vs accent)
- Fixed HTTP method badge showing wrong value for nested/duplicate-titled requests (cmdk selection collision + stale Immer state for nested folder items)
- Known follow-up (not in this PR): same stale-state pattern also affects rename/delete/discard/move for nested items in collectionStore.ts — worth its own issue

## Testing

- Manual tests performed. 
- CommandPalette.test.tsx created

## Checklist

- [X] Issue has been linked to this PR
- [X] Code has been reviewed by person creating the PR
- [X] Automated tests have been written, if possible
- [X] Manual testing has been performed
- [ ] Documentation has been updated, if necessary
- [ ] Changes have been reviewed by second person

___

### **PR Type**
Enhancement

___

### **Description**
- Add Command Palette (Mod+K) with All/Requests/Environments/Actions tabs

- Implement fuzzy search across requests, folders, collections, and actions

- Extract send/save request logic into reusable hooks with shared loading state

- Migrate AppSettingsModal to viewStore-driven open state for palette integration

- Fix hotkey matching to reject unrequested modifiers (e.g., Tab vs Shift+Tab)

- Auto-expand ancestor folders and scroll selected request into view in sidebar

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["User presses Mod+K"] --> B["CommandPalette opens"]
  B --> C["All/Requests/Environments/Actions tabs"]
  C --> D["Fuzzy search results"]
  D --> E["Select request or action"]
  E --> F["Execute action or navigate"]
  F --> G["Palette closes"]
  H["useSendRequest/useSaveRequest hooks"] --> I["Shared loading state"]
  I --> J["MainTopBar and CommandPalette use same logic"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>19 files</summary><table>
<tr>
  <td><strong>App.tsx</strong><dd><code>Integrate CommandPalette and AppSettingsModal into app root</code></dd></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-72348b6b222462c023d009b90544f2799da3cf5b1eafc2d280f8a19041271230">+19/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>CommandPalette.tsx</strong><dd><code>Implement full command palette with tabs and search</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-e722f2b99d46d31fa022681727fcaa973ffde571ef47d3f286ef22cf62ea563b">+458/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>RequestResultsGrid.tsx</strong><dd><code>CSS subgrid layout for aligned request method badges</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-4e03ca31f5946beb3e19c8af301bc12dca8608b3f49c13e574c739a3a3745056">+50/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>MainTopBar.tsx</strong><dd><code>Refactor to use extracted useSendRequest/useSaveRequest hooks</code></dd></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-2ffcaea3cb541bf2a319c21f4b364fa5b0d29c36c183908d44c7f05d4053f9f0">+13/-48</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>InputTabs.tsx</strong><dd><code>Use centralized HOTKEYS constants for tab navigation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href=""></a></td>

</tr>

<tr>
  <td><strong>OutputTabs.tsx</strong><dd><code>Use centralized HOTKEYS constants for response tabs</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-cdb6c46d76626456f074c8fc26e606d739a237d79847a45ca361bc184a0a7be2">+3/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>AppSettingsModal.tsx</strong><dd><code>Convert to controlled component with isOpen and onClose props</code></dd></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-6675a9e18ade8705e607beb1ef50c70fed4b93d22edbb5a98d7a06a9f2de77e7">+8/-14</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>FooterBar.tsx</strong><dd><code>Replace AppSettingsModal with openAppSettings action call</code></dd></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-25c68cd313881a17ee408b9b91fb0801147c9fddda323b67d8721a6dfa241a3a">+4/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>SidebarHeaderBar.tsx</strong><dd><code>Disable mod+n hotkey while command palette is open</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-f309b743e133009344d22d2c557e2a5cfdf6a8c316871c358fb1175dad843a16">+8/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>NavRequest.tsx</strong><dd><code>Auto-scroll selected request into view when highlighted</code>&nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-082424ffb7aaa08e3cf90536823898e941c782b90b82b0d8a65f1606a5613c42">+16/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>SidebarRequestList.tsx</strong><dd><code>Use centralized HOTKEYS constants for request navigation</code>&nbsp; </dd></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-8a3b19e4fb06d66cbab51b9daa33691eb40ae376d75ccba717327eaea8d28007">+3/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>command.tsx</strong><dd><code>Update styling for command palette integration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-a6c50b40e063802c618ab818c48f388b20d0b963f4612b0d1087a7ba07ccae14">+7/-7</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>tabs.tsx</strong><dd><code>Remove fixed height from TabsList for flexible layout</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-2078b9c42084fbef8bb7422fff5864f185a4c22ea86b75a3df7b4798d4f61cde">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>hotkeys.ts</strong><dd><code>Create single source of truth for all global keyboard shortcuts</code></dd></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-f5cf4a2640d322b0265ba36c3c24112a027a75628b8533dc7d16dff3c1f5bc25">+32/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>hotkeyDisplay.tsx</strong><dd><code>Format hotkeys for platform-aware display with Mac/Ctrl badges</code></dd></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-ba8dd411b2cfc0024d6339d22db8129dfdd998f945f85cbb8a66bd725335a8d6">+23/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>useRequestActions.ts</strong><dd><code>Extract send/save request logic with shared loading state</code></dd></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-b0ebdd2e5f5087982512814ec9542e70cb96f2a2071d0993c2e2d9a14fa6c81b">+102/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>collectionStore.ts</strong><dd><code>Auto-expand ancestor folders when selecting nested request</code></dd></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-9bd66bf13fd7c115c613572a46b343298648618f8245cf60b86b6e420d738a32">+12/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>viewStore.ts</strong><dd><code>Add command palette, app settings, and create-item bridge state</code></dd></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-3f3242036ea07710b3828c5ecae6bcf8ed0bdfa1e31e303b9096f46300677a67">+53/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>Menubar.tsx</strong><dd><code>Implement pendingCreateItem bridge from viewStore to sidebar</code></dd></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-cc250f3311590dfd3658ab42da7506282879ccfe6caa66a353c5f7f8a3081085">+15/-1</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>5 files</summary><table>
<tr>
  <td><strong>CommandPalette.test.tsx</strong><dd><code>Comprehensive tests for palette functionality and hotkeys</code></dd></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-5cafa32b0cc2c46ee3d2560d1caec836a0121ca8928674eae7b8e34dae359909">+400/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>SidebarHeaderBar.test.tsx</strong><dd><code>Add tests for hotkey ownership during palette open state</code>&nbsp; </dd></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-5531a71800f40921903cccc5d42f32426a1b4e4deef2a98141e42b94a47a1b34">+39/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>useHotkey.test.ts</strong><dd><code>Add tests for modifier matching edge cases</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-715cae778dcc7c6ee5b0522e1281a6b7acb643617975b365e0d6f74b868911e4">+56/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>useRequestActions.test.ts</strong><dd><code>Test shared loading state across multiple hook instances</code>&nbsp; </dd></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-90fb6d8a927d3aeadd5fab360bbcf718c90ef327bc10b6a59cf4187d7785692b">+147/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>Menubar.test.tsx</strong><dd><code>Test pendingCreateItem bridge and re-fire behavior</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-9ef840e532a16420b317f7d114a4cc23ffd2f56f77ba97a15db35a2f46414776">+104/-0</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Bug_fix</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>useHotkey.ts</strong><dd><code>Fix hotkey matching to reject unrequested modifiers</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-3a6e82627ace060e4872bef9c9cf464ec8dab01675ba1da3541d7e5a69aeb828">+7/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Additional files</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>InputTabs.tsx</strong></td>
  <td><a href="https://github.com/EXXETA/trufos/pull/986/files#diff-34facf6841ac77dbf09ad93791508a9173c180c0700c20440831fc20f6003de3">+6/-5</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: src/renderer/components/commandPalette/CommandPalette.tsx [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+};
+
+const TABS = ['all', 'requests', 'environments', 'actions'] as const;
+type Tab = (typeof TABS)[number];
+
+/** 2-column grid ancestor: column 1 (method badge) auto-sizes to the widest method text across every subgridded row. */
+const REQUEST_LIST_GRID = 'grid grid-cols-[auto_1fr]';
+/** Passes the 2 grid tracks down through cmdk's fixed group/heading/items DOM so every request row's columns line up. */
+const REQUEST_GROUP_GRID =
+  'col-span-full grid grid-cols-subgrid **:[[cmdk-group-heading]]:col-span-full **:[[cmdk-group-items]]:col-span-full **:[[cmdk-group-items]]:grid **:[[cmdk-group-items]]:grid-cols-subgrid';
```
Comment (SoulKa) [inline src/renderer/components/commandPalette/CommandPalette.tsx]: CSS groups in a string constant seem pretty hard to maintain. Can we move this out into a subcomponent?

File: src/renderer/components/commandPalette/CommandPalette.tsx [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+  const collection = useCollectionStore((s) => s.collection);
+  const folders = useCollectionStore((s) => s.folders);
+  const requestGroups = collection ? buildRequestGroups(collection.children, folders) : [];
+  const allRequests = requestGroups.flatMap((group) => group.requests);
+  const currentRequest = useCollectionStore(selectRequest);
+  const { setSelectedRequest, addNewRequest, updateRequest, discardChanges, addNewFolder } =
+    useCollectionActions();
+
+  console.log('allRequests', allRequests);
+  console.log('collection', collection);
```
Comment (SoulKa) [inline src/renderer/components/commandPalette/CommandPalette.tsx]: should not log this

File: src/renderer/components/commandPalette/CommandPalette.tsx [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+    if (currentRequest == null) return;
+    try {
+      await Promise.all(editor.getModels().map(saveModelContent));
+      const response = await httpService.sendRequest(currentRequest);
+      addResponse(currentRequest.id, response);
+    } catch (error) {
+      showError(error);
+    }
+    onClose();
+  }, [currentRequest, addResponse, onClose]);
```
Comment (SoulKa) [inline src/renderer/components/commandPalette/CommandPalette.tsx]: we should not duplicate the sending logic. The save then send logic exists twice now and we had issue with that in earlier versions

File: src/renderer/components/commandPalette/CommandPalette.tsx [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+  const handleSave = useCallback(async () => {
+    if (currentRequest == null) return;
+    try {
+      await Promise.all(editor.getModels().map(saveModelContent));
+      updateRequest(await eventService.saveChanges(currentRequest), true);
+    } catch (error) {
+      showError(error);
+    }
+    onClose();
+  }, [currentRequest, updateRequest, onClose]);
```
Comment (SoulKa) [inline src/renderer/components/commandPalette/CommandPalette.tsx]: same as above

File: src/renderer/components/commandPalette/CommandPalette.tsx [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+    </CommandItem>
+  );
+
+  const actionItems: ActionItem[] = [
+    {
+      value: 'send request',
+      section: 'Request',
+      icon: ArrowRight,
+      label: 'Send request',
+      shortcutModifier: isMac ? <CommandShortcut size={12} /> : 'Ctrl',
```
Comment (SoulKa) [inline src/renderer/components/commandPalette/CommandPalette.tsx]: I think there's a cross platform modifier for this. Check how CMD+S is currently configured

File: src/renderer/components/commandPalette/CommandPalette.tsx [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+      if (e.key === 'ArrowRight' || (e.key === 'Tab' && !e.shiftKey)) {
+        e.preventDefault();
+        setActiveTab(TABS[(currentIndex + 1) % TABS.length]);
+      } else if (e.key === 'ArrowLeft' || (e.key === 'Tab' && e.shiftKey)) {
+        e.preventDefault();
+        setActiveTab(TABS[(currentIndex - 1 + TABS.length) % TABS.length]);
+      }
+    },
+    [activeTab]
+  );
```
Comment (SoulKa) [inline src/renderer/components/commandPalette/CommandPalette.tsx]: we just introduced a hotkey handling library (your PR actually :D). Let's use that

File: src/renderer/components/sidebar/SidebarRequestList/Nav/NavRequest.tsx [hunk 7]
```diff
@@ -18,6 +19,20 @@ export const NavRequest = ({ requestId, depth = 0 }: NavRequestProps) => {
...
+    setNodeRef(node);
+    rowRef.current = node;
+  };
+
+  useEffect(() => {
+    if (isHighlighted) {
+      rowRef.current?.scrollIntoView({ block: 'nearest' });
+    }
+  }, [isHighlighted]);
+
```
Comment (SoulKa) [inline src/renderer/components/sidebar/SidebarRequestList/Nav/NavRequest.tsx:35]: is this really worth having, only to support scroll-to?

**Review by SoulKa — COMMENTED** [review]
Still have to run locally :) only scrolled over the code

**Comment by SoulKa** [comment]
Looks great visually!
image

Only thing I notice that has a weird usage is that creating folder or request will just use random name. Should toggle name input just as it does when pressing button.
Also, request saving is always allowed, even when no changes are made.

File: src/renderer/components/commandPalette/CommandPalette.tsx (hunk 6 above)
Comment (eugen-soliar) [reply src/renderer/components/commandPalette/CommandPalette.tsx]: Not exactly for this case but I'll merge functionality since this is the same responsibility.

File: src/renderer/components/sidebar/SidebarRequestList/Nav/NavRequest.tsx (hunk 7 above)
Comment (eugen-soliar) [reply src/renderer/components/sidebar/SidebarRequestList/Nav/NavRequest.tsx:35]: Otherwise output looks ugly when we have more than a few requests. 
Just imagine you have created a new request with cmd + k palette. And you already have like 60 more. You'd want to see where your new request is (hovered and in the observable area)
So, yes I'd leave it be like it is for UX sake :)

**Review by SoulKa — CHANGES_REQUESTED** [review]
Would approve code wise now, well done, especially the hotkey handling :) only two things are still not fixed:
- Creating folders via action still generates some random name (it is fixed for request now)
- Send request action keeps command palette open, no visual feedback. I would prefer closing the command palette directly, the request view is open already and it would render the send button as waiting indicator -> use sees that request is being sent.

**Review by SoulKa — APPROVED** [review]
nice
